Sources/StreamChat/Repositories/OfflineRequestsRepository.swift

```diff
@@ -0,0 +1,134 @@
+    func runQueuedRequests(completion: @escaping () -> Void) {
+        var pendingActions: [(String, Data)] = []
+
+        let readContext = database.backgroundReadOnlyContext
+        readContext.performAndWait {
+            pendingActions = QueuedRequestDTO.loadAllPendingRequests(context: readContext).map {
+                ($0.id, $0.endpoint)
+            }
+        }
+
+        log.info("\(pendingActions.count) pending offline requests", subsystems: .offlineSupport)
+
+        let database = self.database
+        let group = DispatchGroup()
+        for (id, endpoint) in pendingActions {
+            guard let endpoint = try? JSONDecoder.stream.decode(DataEndpoint.self, from: endpoint) else {
+                continue
+            }
+            log.info("Executing queued offline request for /\(endpoint.path)", subsystems: .offlineSupport)
+
+            group.enter()
+            apiClient.recoveryRequest(endpoint: endpoint) { [weak self] result in
+                let leave = {
+                    group.leave()
+                }
+                let deleteQueuedRequestAndComplete = {
+                    log.info("Completed queued offline request /\(endpoint.path)", subsystems: .offlineSupport)
+                    database.write({ session in
+                        session.deleteQueuedRequest(id: id)
+                    }, completion: { _ in leave() })
+                }
+
+                switch result {
+                case let .success(data):
+                    self?.performDatabaseRecoveryActionsUponSuccess(
+                        for: endpoint,
+                        data: data,
+                        completion: deleteQueuedRequestAndComplete
+                    )
+                case .failure(_ as ClientError.ConnectionError):
+                    // If we failed because there is still no successful connection, we don't remove it from the queue
+                    log.info(
+                        "Keeping offline request /\(endpoint.path) as there is no connection",
+                        subsystems: .offlineSupport
+                    )
+                    leave()
+                case .failure:
+                    deleteQueuedRequestAndComplete()
+                }
+            }
+        }
+
+        group.notify(queue: DispatchQueue.main) {
+            log.info("Done executing all queued offline requests", subsystems: .offlineSupport)
+            completion()
+        }
+    }
+
+    private func performDatabaseRecoveryActionsUponSuccess(
+        for endpoint: DataEndpoint,
+        data: Data,
+        completion: @escaping () -> Void
+    ) {
+        // TODO: Temporary String check. We only process the response for sent messages
+        guard endpoint.path.hasSuffix("/message") else {
+            completion()
+            return
+        }
+        completion()
+    }
+
+    func queueOfflineRequest(endpoint: DataEndpoint, completion: (() -> Void)? = nil) {
```

Comment:
Does it makes sense to provide more info in the completion block? At the moment the caller doesn't know which of the different cases called it.